File: solutions/10SpiralMatrix/python/SpiralMatrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        left, right, top, down = 0, len(matrix[0]) - 1, 0, len(matrix) - 1
        ans = []

        while left <= right & top <= down:
            for j in range(left, right + 1):
                ans.append(matrix[top][j])
            top += 1

            for i in range(top, down + 1):
                ans.append(matrix[i][right])
            right -= 1

            if top <= down:
                for j in range(right, left - 1, -1):
                    ans.append(matrix[down][j])
                down -= 1

            if left <= right:
                for i in range(down, top - 1, -1):
                    ans.append(matrix[i][left])
                left += 1

        return ans

File: solutions/10SpiralMatrix/java/SpiralMatrix.java
import java.util.ArrayList;

public class SpiralMatrix {
  public List<Integer> spiralOrder(int[][] matrix) {
    int rows = matrix.length;
    int cols = matrix[0].length;

    boolean[] colViewed = new boolean[cols];
    boolean[] rowViewed = new boolean[rows];
    boolean right = true, down = false, left = false, up = false;

    int x = 0, y = 0, index = 0;
    List<Integer> answer = new ArrayList<>();

    while (true) {
      answer.add(matrix[x][y]);
      if (right && y < cols - 1 && !colViewed[y + 1]) {
        y += 1;
        continue;
      } else if (right && (y >= cols || colViewed[y + 1]) && (x < rows - 1 && !colViewed[x + 1])) {
        right = false;
        down = true;
        x += 1;
        rowViewed[x] = true;
        continue;
      } else if (down && x < rows - 1 && !rowViewed[x + 1]) {
        x += 1;
        continue;
      } else if (down && (x >= rows || rowViewed[x + 1]) && (y > 0 && !colViewed[y - 1])) {
        down = false;
        left = true;
        x -= 1;
        colViewed[y] = true;
        continue;
      } else if (left && y > 0 && !colViewed[y - 1]) {
        y -= 1;
        continue;
      } else if (left && (y <= 0 || colViewed[y - 1]) && (x > 0 && !rowViewed[x - 1])) {
        left = false;
        up = true;
        x -= 1;
        rowViewed[x] = true;
        continue;
      } else if (up && x > 0 && !rowViewed[x - 1]) {
        x -= 1;
        continue;
      } else if (up && (x <= 0 || rowViewed[x - 1]) && (y < cols - 1 && !colViewed[y + 1])) {
        up = false;
        right = true;
        y += 1;
        colViewed[y] = true;
        continue;
      } else {
        return answer;
      }
    }
  }
}
